### PR TITLE
Feature/follow point

### PR DESCRIPTION
**Pull request by rickaha**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Just a cleanup to separate follow point from waypointfollower

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

Same as before, no change

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:
Has been tested with DepthAI camera, both follow point and the emergency brake function when driving a route.
Build tests for RCCar can't be passed before merging following pull request: https://github.com/RISE-Dependable-Transport-Systems/RCCar/pull/13

### CODE REVIEW

File: autopilot/followpoint.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    emit deactivateEmergencyBrake();
+    if (isOnVehicle()) {
+        mStmState = FollowPointSTMstates::FOLLOWING;
+        mUpdateStateTimer.start(mUpdateStatePeriod_ms);
+    } else {
+        if (isOnVehicle())
+            qDebug() << "WARNING: Follow Point did not get a recent point to follow. Exiting.";
+        else
+            qDebug() << "WARNING: Follow Point not implemented for remote connections. Exiting.";
+        mStmState = FollowPointSTMstates::STOP;
```
Comment (marvind) [inline autopilot/followpoint.cpp:56]: Would add mUpdateStateTimer.stop() as well to make sure the STM is not executed in this state.

File: autopilot/followpoint.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        return mVehicleConnection->getVehicleState()->getPosition(mPosTypeUsed);
+}
+
+void FollowPoint::updateState()
+{
+    QLineF egoVehicleToFollowPointLine(QPointF(0,0), mCurrentFollowPointInVehicleFrame.getPoint());
+
+    switch (mStmState) {
+    case FollowPointSTMstates::STOP:
+        break;
```
Comment (marvind) [inline autopilot/followpoint.cpp:112]:
```suggestion
    case FollowPointSTMstates::NONE:
        qDebug() << "WARNING: FollowPoint running uninitialized statemachine.";
        break;
```
To make sure you get a warning when the STM state is executed, which to my understanding should never happen. Would call it NONE for consistency with waypointfollower

File: autopilot/followpoint.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#include <QTimer>
+#include <QSharedPointer>
+#include <QPointF>
+#include "core/pospoint.h"
+#include "vehicles/controller/movementcontroller.h"
+#include "communication/vehicleconnections/vehicleconnection.h"
+
+enum class FollowPointSTMstates {STOP, FOLLOWING, WAITING};
+
+class FollowPoint : public QObject
```
Comment (marvind) [inline autopilot/followpoint.h:21]: Hmmm should this be called "PointFollower" for consistency? 🤔

File: communication/mavsdkvehicleserver.cpp [hunk 4]
```diff
@@ -98,9 +98,12 @@ MavsdkVehicleServer::MavsdkVehicleServer(QSharedPointer<VehicleState> vehicleSta
         }
 
         if (mode != mavsdk::ActionServer::FlightMode::Mission &&
-                mode != mavsdk::ActionServer::FlightMode::Offboard)
-            if (mWaypointFollower->isActive())
+                mode != mavsdk::ActionServer::FlightMode::Offboard) {
+            if (!mWaypointFollower.isNull() && mWaypointFollower->isActive())
                 emit pauseWaypointFollower();
+            if (!mFollowPoint.isNull() && mWaypointFollower->isActive())
```
Comment (marvind) [inline communication/mavsdkvehicleserver.cpp:104]:
```suggestion
            if (!mFollowPoint.isNull() && mFollowPoint->isActive())
```

I think the whole part should actually be:
```
        if (mode != mavsdk::ActionServer::FlightMode::Mission)
            if (!mWaypointFollower.isNull() && mWaypointFollower->isActive())
                emit pauseWaypointFollower();

        if (mode != mavsdk::ActionServer::FlightMode::Offboard)
            if (!mFollowPoint.isNull() && mWaypointFollower->isActive())
                emit stopFollowPoint();
```
-> WPF should only be able to run in mission mode, FP only in Offboard mode, right?

File: communication/mavsdkvehicleserver.cpp [hunk 5]
```diff
@@ -498,14 +501,20 @@ void MavsdkVehicleServer::setWaypointFollower(QSharedPointer<WaypointFollower> w
...
 
 void MavsdkVehicleServer::setMovementController(QSharedPointer<MovementController> movementController)
 {
     mMovementController = movementController;
 }
 
+void MavsdkVehicleServer::setFollowPoint(QSharedPointer<FollowPoint> followPoint)
+{
+    mFollowPoint = followPoint;
+    connect(this, &MavsdkVehicleServer::startFollowPoint, mFollowPoint.get(), &FollowPoint::startFollowPoint);
```
Comment (marvind) [inline communication/mavsdkvehicleserver.cpp:514]: Related to the previous comment: previously, WPF and FP had a  single STM that made sure that both autopilots could not be active at the same time. How is this now ensured? Is the change above sufficient or will MavsdkVehicleServer::startFollowPoint need to call sth like (and vice verse for WPF)
```
if (!mWaypointFollower.isNull() && mWaypointFollower->isActive())
                emit pauseWaypointFollower();
```

**Review by marvind — CHANGES_REQUESTED** [review]
Great to see this finally done. 😊 I have a few thoughts I would like to see clarified before merging.

File: communication/iso22133vehicleserver.cpp [hunk 6]
```diff
@@ -216,3 +216,8 @@ void iso22133VehicleServer::updateRawGpsAndGpsInfoFromUbx(
     const ubx_nav_pvt &pvt) {
     // Not implemented
 }
+
+void iso22133VehicleServer::setFollowPoint(
+    QSharedPointer<FollowPoint> followPoint) {
+    // Not implemented
```
[the remark this answers is not in the document]
Comment (rickaha) [reply communication/iso22133vehicleserver.cpp:222]: Agree

File: autopilot/followpoint.cpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+void FollowPoint::startFollowPoint()
+{
+    // Deactivate emergency brake in order to use follow point
+    emit deactivateEmergencyBrake();
+    if (isOnVehicle()) {
+        mStmState = FollowPointSTMstates::FOLLOWING;
+        mUpdateStateTimer.start(mUpdateStatePeriod_ms);
+    } else {
+        if (isOnVehicle())
+            qDebug() << "WARNING: Follow Point did not get a recent point to follow. Exiting.";
```
[the remark this answers is not in the document]
Comment (rickaha) [reply autopilot/followpoint.cpp:53]: I will go through it again and look for bugs. Implementing support for VehicleConnections is the next step, correct.

File: communication/mavsdkvehicleserver.cpp (hunk 4 above)
Comment (rickaha) [reply communication/mavsdkvehicleserver.cpp:104]: Correct

File: communication/mavsdkvehicleserver.cpp (hunk 5 above)
Comment (rickaha) [reply communication/mavsdkvehicleserver.cpp:514]: The way to change mode is through the above function, which with the suggested change, will halt the other follower. I think it should be sufficient for now. This check shouldn't need to be performed for each communication interface, so a common STM is a better solution...

**Pull request merged**